File: api/apps/knative.py
from __future__ import annotations

import json
import os
import ssl
import hashlib
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode
from urllib.request import Request, urlopen


def env(key: str, fallback: str) -> str:
    value = os.getenv(key)
    return value if value else fallback


def now() -> str:
    return datetime.now(tz=timezone.utc).isoformat().replace("+00:00", "Z")


def is_dns_label(value: str) -> bool:
    if len(value) == 0 or len(value) > 63:
        return False
    if value[0] == "-" or value[-1] == "-":
        return False
    for ch in value:
        if ("a" <= ch <= "z") or ("0" <= ch <= "9") or ch == "-":
            continue
        return False
    return True


def public_service_url(name: str, public_domain: str) -> str:
    return f"https://{name}.{public_domain}"


def sanitize_dns_label(value: str) -> str:
    value = value.strip().lower()
    chars: list[str] = []
    last_hyphen = False
    for ch in value:
        if ("a" <= ch <= "z") or ("0" <= ch <= "9"):
            chars.append(ch)
            last_hyphen = False
        elif chars and not last_hyphen:
            chars.append("-")
            last_hyphen = True
    return "".join(chars).strip("-")


def service_resource_name(project_id: str, name: str) -> str:
    seed = f"{project_id.strip()}:{name.strip()}"
    digest = hashlib.sha256(seed.encode("utf-8")).hexdigest()[:8]
    prefix = sanitize_dns_label(name)
    if not prefix:
        prefix = "service"
    max_prefix_len = 63 - 1 - len(digest)
    if len(prefix) > max_prefix_len:
        prefix = prefix[:max_prefix_len].rstrip("-")
    if not prefix:
        prefix = "service"
    return f"{prefix}-{digest}"


@dataclass
class KnativeService:
    name: str
    resource_name: str
    image: str
    url: str
    ready: bool
    reason: str
    created_at: str
    updated_at: str
    namespace: str
    generation: int


@dataclass
class KnativeManager:
    namespace: str
    public_domain: str
    base_url: str
    token: str
    ssl_context: ssl.SSLContext

    @classmethod
    def new(cls, namespace: str, public_domain: str) -> "KnativeManager":
        base_url = f"https://{env('KUBERNETES_SERVICE_HOST', 'kubernetes.default.svc')}"
        token_path = env("DCLD_KUBERNETES_TOKEN_FILE", "/var/run/secrets/kubernetes.io/serviceaccount/token")
        ca_path = env("DCLD_KUBERNETES_CA_FILE", "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt")

        with open(token_path, "r", encoding="utf-8") as handle:
            token = handle.read().strip()

        context = ssl.create_default_context()
        if os.path.exists(ca_path):
            context = ssl.create_default_context(cafile=ca_path)

        return cls(
            namespace=namespace,
            public_domain=public_domain,
            base_url=base_url,
            token=token,
            ssl_context=context,
        )

    def public_url(self, name: str) -> str:
        return public_service_url(name, self.public_domain)

    def list_services(self, project_id: str) -> list[KnativeService]:
        selector = urlencode({"labelSelector": f"dcloud.dev/project={project_id}"})
        payload = self._request_json(
            "GET",
            f"/apis/serving.knative.dev/v1/namespaces/{self.namespace}/services?{selector}",
        )

        items = payload.get("items", [])
        services: list[KnativeService] = []
        for item in items:
            metadata = item.get("metadata", {})
            spec = item.get("spec", {})
            status = item.get("status", {})
            containers = (((spec.get("template") or {}).get("spec") or {}).get("containers")) or []
            reason = ""
            ready = False
            updated_at = metadata.get("creationTimestamp", "")
            for condition in status.get("conditions", []):
                if condition.get("type") == "Ready":
                    ready = condition.get("status") == "True"
                    reason = condition.get("reason") or ""
                    updated_at = condition.get("lastTransitionTime") or updated_at
                    break
            services.append(
                KnativeService(
                    name=metadata.get("name", ""),
                    resource_name=metadata.get("name", ""),
                    image=containers[0].get("image", "") if containers else "",
                    url=self.public_url(metadata.get("name", "")),
                    ready=ready,
                    reason=reason,
                    created_at=metadata.get("creationTimestamp", ""),
                    updated_at=updated_at,
                    namespace=metadata.get("namespace", self.namespace),
                    generation=int(metadata.get("generation") or 0),
                )
            )
        return services

    def deploy_service(self, project_id: str, name: str, image: str, port: int) -> KnativeService:
        if not is_dns_label(name):
            raise ValueError("name must be a DNS label")
        if port < 1 or port > 65535:
            raise ValueError("port must be between 1 and 65535")

        timestamp = now()
        resource_name = service_resource_name(project_id, name)
        manifest = {
            "apiVersion": "serving.knative.dev/v1",
            "kind": "Service",
            "metadata": {
                "name": resource_name,
                "namespace": self.namespace,
                "labels": {
                    "app.kubernetes.io/instance": "dcloud",
                    "app.kubernetes.io/component": "container",
                    "dcloud.dev/project": project_id,
                    "dcloud.dev/service-name": name,
                    "app.kubernetes.io/managed-by": "dcloud-api",
                },
            },
            "spec": {
                "template": {
                    "metadata": {
                        "labels": {
                            "app.kubernetes.io/instance": "dcloud",
                            "app.kubernetes.io/component": "container",
                            "dcloud.dev/project": project_id,
                            "dcloud.dev/service-name": name,
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "name": name,
                                "image": image,
                                "ports": [{"containerPort": port}],
                            }
                        ]
                    },
                }
            },
        }
        service_payload = self._request_json(
            "PATCH",
            f"/apis/serving.knative.dev/v1/namespaces/{self.namespace}/services/{resource_name}?fieldManager=dcloud-api&force=true",
            body=manifest,
            content_type="application/apply-patch+yaml",
        )

        domain_mapping = {
            "apiVersion": "serving.knative.dev/v1beta1",
            "kind": "DomainMapping",
            "metadata": {
                "name": f"{resource_name}.{self.public_domain}",
                "namespace": self.namespace,
                "labels": {
                    "app.kubernetes.io/instance": "dcloud",
                    "app.kubernetes.io/component": "container",
                    "dcloud.dev/project": project_id,
                    "dcloud.dev/service-name": name,
                    "app.kubernetes.io/managed-by": "dcloud-api",
                },
            },
            "spec": {
                "ref": {
                    "apiVersion": "serving.knative.dev/v1",
                    "kind": "Service",
                    "name": resource_name,
                }
            },
        }
        self._request_json(
            "PATCH",
            f"/apis/serving.knative.dev/v1beta1/namespaces/{self.namespace}/domainmappings/{resource_name}.{self.public_domain}?fieldManager=dcloud-api&force=true",
            body=domain_mapping,
            content_type="application/apply-patch+yaml",
        )

        metadata = service_payload.get("metadata", {})
        spec = service_payload.get("spec", {})
        status = service_payload.get("status", {})
        containers = (((spec.get("template") or {}).get("spec") or {}).get("containers")) or []
        ready = False
        reason = ""
        updated_at = timestamp
        for condition in status.get("conditions", []):
            if condition.get("type") == "Ready":
                ready = condition.get("status") == "True"
                reason = condition.get("reason") or ""
                updated_at = condition.get("lastTransitionTime") or timestamp
                break

        return KnativeService(
            name=name,
            resource_name=metadata.get("name", resource_name),
            image=containers[0].get("image", image) if containers else image,
            url=self.public_url(resource_name),
            ready=ready,
            reason=reason,
            created_at=metadata.get("creationTimestamp", timestamp),
            updated_at=updated_at,
            namespace=metadata.get("namespace", self.namespace),
            generation=int(metadata.get("generation") or 1),
        )

    def delete_service(self, project_id: str, name: str) -> None:
        resource_name = service_resource_name(project_id, name)
        self._request_json(
            "DELETE",
            f"/apis/serving.knative.dev/v1beta1/namespaces/{self.namespace}/domainmappings/{resource_name}.{self.public_domain}",
            allow_not_found=True,
        )
        self._request_json(
            "DELETE",
            f"/apis/serving.knative.dev/v1/namespaces/{self.namespace}/services/{resource_name}",
            allow_not_found=True,
        )

    def _request_json(
        self,
        method: str,
        path: str,
        *,
        body: dict[str, Any] | None = None,
        content_type: str = "application/json",
        allow_not_found: bool = False,
    ) -> dict[str, Any]:
        request = Request(self.base_url + path, method=method)
        request.add_header("Authorization", f"Bearer {self.token}")
        request.add_header("Accept", "application/json")
        if body is not None:
            request.add_header("Content-Type", content_type)
            request.data = json.dumps(body).encode("utf-8")

        try:
            with urlopen(request, context=self.ssl_context, timeout=20) as response:
                raw = response.read()
        except HTTPError as exc:
            if allow_not_found and exc.code == 404:
                return {}
            raw = exc.read()
            message = self._decode_error(raw)
            raise RuntimeError(message or f"kubernetes api returned {exc.code}") from exc
        except URLError as exc:
            raise RuntimeError(f"failed to reach kubernetes api: {exc.reason}") from exc

        if not raw:
            return {}
        return json.loads(raw.decode("utf-8"))

    @staticmethod
    def _decode_error(raw: bytes) -> str:
        if not raw:
            return ""
        try:
            payload = json.loads(raw.decode("utf-8"))
        except json.JSONDecodeError:
            return ""
        message = payload.get("message") or payload.get("reason")
        return str(message) if message else ""

File: internal/container/cmd/server/knative.go
package main

import (
	"bytes"
	"context"
	"crypto/sha256"
	"crypto/tls"
	"crypto/x509"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"strings"
	"time"
)

const (
	userServiceManagerLabel   = "dcloud-container"
	userLabelKey              = "dcloud.dev/user"
	projectLabelKey           = "dcloud.dev/project"
	serviceNameLabel          = "dcloud.dev/service-name"
)

var errKnativeServiceNotFound = errors.New("service not found")

type knativeServiceManager struct {
	namespace    string
	publicDomain string
	client       *http.Client
	baseURL      string
	token        string
}

func newKnativeServiceManager(namespace string, publicDomain string) (*knativeServiceManager, error) {
	baseURL := fmt.Sprintf("https://%s", env("KUBERNETES_SERVICE_HOST", "kubernetes.default.svc"))
	tokenPath := env("DCLD_KUBERNETES_TOKEN_FILE", "/var/run/secrets/kubernetes.io/serviceaccount/token")
	caPath := env("DCLD_KUBERNETES_CA_FILE", "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt")

	tokenBytes, err := os.ReadFile(tokenPath)
	if err != nil {
		return nil, err
	}

	rootCAs, err := x509.SystemCertPool()
	if err != nil || rootCAs == nil {
		rootCAs = x509.NewCertPool()
	}
	if caBytes, readErr := os.ReadFile(caPath); readErr == nil {
		rootCAs.AppendCertsFromPEM(caBytes)
	}

	return &knativeServiceManager{
		namespace:    namespace,
		publicDomain: publicDomain,
		baseURL:      baseURL,
		token:        strings.TrimSpace(string(tokenBytes)),
		client: &http.Client{
			Timeout: 20 * time.Second,
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{RootCAs: rootCAs},
			},
		},
	}, nil
}

func (m *knativeServiceManager) publicURL(resourceName string) string {
	return fmt.Sprintf("https://%s.%s", resourceName, m.publicDomain)
}

func (m *knativeServiceManager) customURL(domain string) string {
	return fmt.Sprintf("https://%s", domain)
}

func (m *knativeServiceManager) applyDomainMapping(ctx context.Context, domainName, resourceName string, labels map[string]string) error {
	body, err := json.Marshal(map[string]any{
		"apiVersion": "serving.knative.dev/v1beta1",
		"kind":       "DomainMapping",
		"metadata": map[string]any{
			"name":      domainName,
			"namespace": m.namespace,
			"labels":    labels,
		},
		"spec": map[string]any{
			"ref": map[string]any{
				"apiVersion": "serving.knative.dev/v1",
				"kind":       "Service",
				"name":       resourceName,
			},
		},
	})
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPatch,
		fmt.Sprintf("%s/apis/serving.knative.dev/v1beta1/namespaces/%s/domainmappings/%s?fieldManager=dcloud-container&force=true", m.baseURL, m.namespace, domainName),
		bytes.NewReader(body))
	if err != nil {
		return err
	}
	m.authorize(req)
	req.Header.Set("Content-Type", "application/apply-patch+yaml")
	req.Header.Set("Accept", "application/json")
	res, err := m.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode >= 300 {
		return decodeAPIError(res)
	}
	return nil
}

func (m *knativeServiceManager) deleteDomainMapping(ctx context.Context, domainName string) error {
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete,
		fmt.Sprintf("%s/apis/serving.knative.dev/v1beta1/namespaces/%s/domainmappings/%s", m.baseURL, m.namespace, domainName),
		nil)
	if err != nil {
		return err
	}
	m.authorize(req)
	res, err := m.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode >= 300 && res.StatusCode != http.StatusNotFound {
		return decodeAPIError(res)
	}
	return nil
}

func (m *knativeServiceManager) setCustomDomain(ctx context.Context, scope projectScope, name, customDomain string) error {
	resourceName := serviceResourceName(scope.ProjectID, name)
	labels := map[string]string{
		"app.kubernetes.io/instance":   "dcloud",
		"app.kubernetes.io/component":  "container",
		"app.kubernetes.io/managed-by": userServiceManagerLabel,
		userLabelKey:                   scope.UserID,
		projectLabelKey:                scope.ProjectID,
		serviceNameLabel:               name,
	}
	return m.applyDomainMapping(ctx, customDomain, resourceName, labels)
}

func (m *knativeServiceManager) list(ctx context.Context, scope projectScope) ([]deployedService, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("%s/apis/serving.knative.dev/v1/namespaces/%s/services?labelSelector=%s", m.baseURL, m.namespace, strings.Join([]string{
		projectLabelKey + "=" + strings.TrimSpace(scope.ProjectID),
	}, ",")), nil)
	if err != nil {
		return nil, err
	}
	m.authorize(req)

	res, err := m.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer res.Body.Close()
	if res.StatusCode >= 300 {
		return nil, decodeAPIError(res)
	}

	var payload struct {
		Items []struct {
			Metadata struct {
				Name              string            `json:"name"`
				CreationTimestamp time.Time         `json:"creationTimestamp"`
				Generation        int64             `json:"generation"`
				Namespace         string            `json:"namespace"`
				Labels            map[string]string `json:"labels"`
			} `json:"metadata"`
			Spec struct {
				Template struct {
					Spec struct {
						Containers []struct {
							Image string `json:"image"`
						} `json:"containers"`
					} `json:"spec"`
				} `json:"template"`
			} `json:"spec"`
			Status struct {
				Conditions []struct {
					Type               string    `json:"type"`
					Status             string    `json:"status"`
					Reason             string    `json:"reason"`
					LastTransitionTime time.Time `json:"lastTransitionTime"`
				} `json:"conditions"`
			} `json:"status"`
		} `json:"items"`
	}
	if err := json.NewDecoder(res.Body).Decode(&payload); err != nil {
		return nil, err
	}

	services := make([]deployedService, 0, len(payload.Items))
	for _, item := range payload.Items {
		displayName := strings.TrimSpace(item.Metadata.Labels[serviceNameLabel])
		if displayName == "" {
			displayName = item.Metadata.Name
		}
		svc := deployedService{
			Name:         displayName,
			Image:        "",
			URL:          m.publicURL(item.Metadata.Name),
			ResourceName: item.Metadata.Name,
			CreatedAt:    item.Metadata.CreationTimestamp.UTC().Format(time.RFC3339),
			UpdatedAt:    item.Metadata.CreationTimestamp.UTC().Format(time.RFC3339),
			Namespace:    item.Metadata.Namespace,
			ProjectID:    item.Metadata.Labels[projectLabelKey],
			Generation:   item.Metadata.Generation,
		}
		if len(item.Spec.Template.Spec.Containers) > 0 {
			svc.Image = item.Spec.Template.Spec.Containers[0].Image
		}
		for _, cond := range item.Status.Conditions {
			if cond.Type == "Ready" {
				svc.Ready = cond.Status == "True"
				svc.Reason = cond.Reason
				if !cond.LastTransitionTime.IsZero() {
					svc.UpdatedAt = cond.LastTransitionTime.UTC().Format(time.RFC3339)
				}
				break
			}
		}
		services = append(services, svc)
	}
	return services, nil
}

func (m *knativeServiceManager) deploy(ctx context.Context, scope projectScope, req deployRequest) (deployedService, error) {
	resourceName := serviceResourceName(scope.ProjectID, req.Name)
	manifest := knativeServiceManifest{
		APIVersion: "serving.knative.dev/v1",
		Kind:       "Service",
	}
	manifest.Metadata.Name = resourceName
	manifest.Metadata.Namespace = m.namespace
	manifest.Metadata.Labels = map[string]string{
		"app.kubernetes.io/instance":   "dcloud",
		"app.kubernetes.io/component":  "container",
		"app.kubernetes.io/managed-by": userServiceManagerLabel,
		userLabelKey:                   scope.UserID,
		projectLabelKey:                scope.ProjectID,
		serviceNameLabel:               req.Name,
	}
	manifest.Spec.Template.Metadata.Labels = map[string]string{
		"app.kubernetes.io/instance":  "dcloud",
		"app.kubernetes.io/component": "container",
		userLabelKey:                  scope.UserID,
		projectLabelKey:               scope.ProjectID,
		serviceNameLabel:              req.Name,
	}
	if req.MinScale > 0 || req.MaxScale > 0 {
		manifest.Spec.Template.Metadata.Annotations = map[string]string{}
		if req.MinScale > 0 {
			manifest.Spec.Template.Metadata.Annotations["autoscaling.knative.dev/minScale"] = fmt.Sprintf("%d", req.MinScale)
		}
		if req.MaxScale > 0 {
			manifest.Spec.Template.Metadata.Annotations["autoscaling.knative.dev/maxScale"] = fmt.Sprintf("%d", req.MaxScale)
		}
	}
	manifest.Spec.Template.Spec.Containers = []knativeContainer{{
		Name:  req.Name,
		Image: req.Image,
		Ports: []knativeContainerPort{{ContainerPort: req.Port}},
	}}

	body, err := json.Marshal(manifest)
	if err != nil {
		return deployedService{}, err
	}
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPatch, fmt.Sprintf("%s/apis/serving.knative.dev/v1/namespaces/%s/services/%s?fieldManager=dcloud-container&force=true", m.baseURL, m.namespace, resourceName), bytes.NewReader(body))
	if err != nil {
		return deployedService{}, err
	}
	m.authorize(httpReq)
	httpReq.Header.Set("Content-Type", "application/apply-patch+yaml")
	httpReq.Header.Set("Accept", "application/json")

	res, err := m.client.Do(httpReq)
	if err != nil {
		return deployedService{}, err
	}
	defer res.Body.Close()
	if res.StatusCode >= 300 {
		return deployedService{}, decodeAPIError(res)
	}

	var payload struct {
		Metadata struct {
			Name              string    `json:"name"`
			CreationTimestamp time.Time `json:"creationTimestamp"`
			Generation        int64     `json:"generation"`
			Namespace         string    `json:"namespace"`
		} `json:"metadata"`
		Spec struct {
			Template struct {
				Spec struct {
					Containers []struct {
						Image string `json:"image"`
					} `json:"containers"`
				} `json:"spec"`
			} `json:"template"`
		} `json:"spec"`
		Status struct {
			Conditions []struct {
				Type               string    `json:"type"`
				Status             string    `json:"status"`
				Reason             string    `json:"reason"`
				LastTransitionTime time.Time `json:"lastTransitionTime"`
			} `json:"conditions"`
		} `json:"status"`
	}
	if err := json.NewDecoder(res.Body).Decode(&payload); err != nil {
		return deployedService{}, err
	}

	defaultDomainLabels := map[string]string{
		"app.kubernetes.io/instance":   "dcloud",
		"app.kubernetes.io/component":  "container",
		"app.kubernetes.io/managed-by": userServiceManagerLabel,
		userLabelKey:                   scope.UserID,
		projectLabelKey:                scope.ProjectID,
		serviceNameLabel:               req.Name,
	}
	if err := m.applyDomainMapping(ctx, fmt.Sprintf("%s.%s", resourceName, m.publicDomain), resourceName, defaultDomainLabels); err != nil {
		return deployedService{}, err
	}

	service := deployedService{
		Name:         req.Name,
		Image:        req.Image,
		URL:          m.publicURL(resourceName),
		ResourceName: resourceName,
		Namespace:    payload.Metadata.Namespace,
		ProjectID:    scope.ProjectID,
		Generation:   payload.Metadata.Generation,
		CreatedAt:    payload.Metadata.CreationTimestamp.UTC().Format(time.RFC3339),
		UpdatedAt:    payload.Metadata.CreationTimestamp.UTC().Format(time.RFC3339),
	}
	if len(payload.Spec.Template.Spec.Containers) > 0 {
		service.Image = payload.Spec.Template.Spec.Containers[0].Image
	}
	for _, cond := range payload.Status.Conditions {
		if cond.Type == "Ready" {
			service.Ready = cond.Status == "True"
			service.Reason = cond.Reason
			if !cond.LastTransitionTime.IsZero() {
				service.UpdatedAt = cond.LastTransitionTime.UTC().Format(time.RFC3339)
			}
			break
		}
	}
	return service, nil
}

func (m *knativeServiceManager) delete(ctx context.Context, scope projectScope, name, customDomain string) error {
	resourceName := serviceResourceName(scope.ProjectID, name)
	if customDomain != "" {
		if err := m.deleteDomainMapping(ctx, customDomain); err != nil {
			return err
		}
	}
	if err := m.deleteDomainMapping(ctx, fmt.Sprintf("%s.%s", resourceName, m.publicDomain)); err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodDelete, fmt.Sprintf("%s/apis/serving.knative.dev/v1/namespaces/%s/services/%s", m.baseURL, m.namespace, resourceName), nil)
	if err != nil {
		return err
	}
	m.authorize(req)
	res, err := m.client.Do(req)
	if err != nil {
		return err
	}
	defer res.Body.Close()
	if res.StatusCode >= 300 && res.StatusCode != http.StatusNotFound {
		return decodeAPIError(res)
	}
	return nil
}

func (m *knativeServiceManager) authorize(req *http.Request) {
	req.Header.Set("Authorization", "Bearer "+m.token)
}

func serviceResourceName(projectID, name string) string {
	seed := strings.TrimSpace(projectID) + ":" + strings.TrimSpace(name)
	sum := sha256.Sum256([]byte(seed))
	suffix := hex.EncodeToString(sum[:4])
	prefix := sanitizeDNSLabel(name)
	if prefix == "" {
		prefix = "service"
	}
	maxPrefixLen := 63 - 1 - len(suffix)
	if len(prefix) > maxPrefixLen {
		prefix = strings.TrimRight(prefix[:maxPrefixLen], "-")
	}
	if prefix == "" {
		prefix = "service"
	}
	return prefix + "-" + suffix
}

func sanitizeDNSLabel(value string) string {
	value = strings.ToLower(strings.TrimSpace(value))
	var builder strings.Builder
	lastHyphen := false
	for _, ch := range value {
		switch {
		case ch >= 'a' && ch <= 'z':
			builder.WriteRune(ch)
			lastHyphen = false
		case ch >= '0' && ch <= '9':
			builder.WriteRune(ch)
			lastHyphen = false
		case builder.Len() > 0 && !lastHyphen:
			builder.WriteRune('-')
			lastHyphen = true
		}
	}
	return strings.Trim(builder.String(), "-")
}

type knativeServiceManifest struct {
	APIVersion string `json:"apiVersion"`
	Kind       string `json:"kind"`
	Metadata   struct {
		Name      string            `json:"name"`
		Namespace string            `json:"namespace"`
		Labels    map[string]string `json:"labels"`
	} `json:"metadata"`
	Spec struct {
		Template struct {
			Metadata struct {
				Labels      map[string]string `json:"labels"`
				Annotations map[string]string `json:"annotations,omitempty"`
			} `json:"metadata"`
			Spec struct {
				Containers []knativeContainer `json:"containers"`
			} `json:"spec"`
		} `json:"template"`
	} `json:"spec"`
}

type knativeContainer struct {
	Name  string               `json:"name"`
	Image string               `json:"image"`
	Ports []knativeContainerPort `json:"ports"`
}

type knativeContainerPort struct {
	ContainerPort int32 `json:"containerPort"`
}

func decodeAPIError(res *http.Response) error {
	var payload struct {
		Message string `json:"message"`
		Reason  string `json:"reason"`
		Code    int    `json:"code"`
	}
	if err := json.NewDecoder(res.Body).Decode(&payload); err == nil {
		if payload.Message != "" {
			return fmt.Errorf("%s", payload.Message)
		}
		if payload.Reason != "" {
			return fmt.Errorf("%s", payload.Reason)
		}
	}
	return fmt.Errorf("kubernetes api returned %s", res.Status)
}
